contracts/validator-manager/PoSValidatorManager.sol

```diff
@@ -486,46 +424,63 @@ abstract contract PoSValidatorManager is
         // Set the delegator status to pending removed, so that it can be properly removed in
         // the complete step, even if the delivered nonce is greater than the nonce used to
         // initialize the removal.
-        delegator.status = DelegatorStatus.PendingRemoved;
+        $._delegatorStakes[delegationID].status = DelegatorStatus.PendingRemoved;
 
-        uint64 validatorUptimeSeconds;
-        uint64 delegationEndTime;
         if (validator.status == ValidatorStatus.Active) {
             if (includeUptimeProof) {
                 // Uptime proofs include the absolute number of seconds the validator has been active.
-                validatorUptimeSeconds = _getUptime(validationID, messageIndex);
+                $._completedValidationUptimeSeconds[validationID] =
+                    _getUptime(validationID, messageIndex);
             }
-            uint64 newValidatorWeight = validator.weight - delegator.weight;
-            (delegator.endingNonce,) = _setValidatorWeight(validationID, newValidatorWeight);
 
-            delegationEndTime = uint64(block.timestamp);
+            ($._delegatorStakes[delegationID].endingNonce,) =
+                _setValidatorWeight(validationID, validator.weight - delegator.weight);
+            $._redeemableDelegatorRewards[delegationID] =
+                _calculateDelegationReward(delegator, validator);
+        } else if (validator.status == ValidatorStatus.Completed) {
+            $._redeemableDelegatorRewards[delegationID] =
+                _calculateDelegationReward(delegator, validator);
+            return _completeEndDelegation(delegationID);
         } else {
-            // If the validation period has already ended, we have saved the uptime.
-            // Further, it is impossible to retrieve an uptime proof for an already ended validation,
-            // so there's no need to check any uptime proof provided in this function call.
-            validatorUptimeSeconds = $._completedValidationUptimeSeconds[validationID];
+            revert InvalidValidatorStatus();
+        }
 
-            delegator.endingNonce = validator.messageNonce;
+        emit DelegatorRemovalInitialized({delegationID: delegationID, validationID: validationID});
```

Comment:
nit: perhaps move this to the end of the "if" block since both the "else if" and "else" blocks either return or revert and doesn't get here.